onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,218 @@
+    auto& input_dim = input_shape->dim(input_rank - 1 - i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_value() == input_dim.dim_value());
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_param() == input_dim.dim_param());
+    }
+  }
+
+  return SwapGatherNDWithInputNode(graph, gathernd_node, gathernd_input_node, non_weight_input_index);
+}
+
+/*
+  This handler change the graphs this way:
+    Before:
+      input_1[b,s,h]    weight_2[h, 2h]
+                  \         /
+                  MatMul[b,s,2h]    indices[b,p_s,1]
+                      |            /
+                  GatherND[b,p_s,2h]
+                      |
+                <subsquent graphs>
+    After :
+               input_1[b,s,h]      indices[b,p_s,1]
+                       |            /
+                    GatherND[b,p_s,h]    weight_2[h,2h]
+                              \              /
+                              Add[b,p_s,2h]
+                                    |
+                            <subsquent graphs>
+  Note: b: batch, s: sequence_length, h: hidden_size, p_s: dynamic_prediction_count
+*/
+static Status MatMulHandler(Graph& graph, Node* gathernd_node, Node* gathernd_input_node) {
+  int non_weight_input_index = -1;
+  for (size_t i = 0; i < gathernd_input_node->MutableInputDefs().size(); i++) {
+    if (!graph_utils::IsGraphInput(graph, gathernd_input_node->MutableInputDefs()[i]) &&
+        !graph_utils::IsInitializer(graph, gathernd_input_node->MutableInputDefs()[i]->Name(), false)) {
+      non_weight_input_index = i;
+    }
+  }
+
+  ORT_RETURN_IF_NOT(non_weight_input_index == 0);
+  auto output_shape = gathernd_input_node->MutableOutputDefs()[0]->Shape();
+  auto input_shape = gathernd_input_node->MutableInputDefs()[non_weight_input_index]->Shape();
+
+  ORT_RETURN_IF_NOT(output_shape != nullptr && input_shape != nullptr);
+  const int output_rank = output_shape->dim_size();
+  const int input_rank = input_shape->dim_size();
+  ORT_RETURN_IF_NOT(input_rank == output_rank);
+  // Only hande cases where input share exactly equal with output shape on the first two dims (
+  // because GatherND's batch_dims be 1).
+  // otherwise, it might not be safe to move GatherND upward.
+  for (int i = 0; i < 2; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_value() == input_dim.dim_value());
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_param() == input_dim.dim_param());
+    }
+  }
+  return SwapGatherNDWithInputNode(graph, gathernd_node, gathernd_input_node, 0);
+}
+
+static std::unordered_map<std::string, Handler> handlers = {
+    {"Add", AddHandler},
+    {"Gelu", SimpleHandler},
```

Comment:
Currently the answer is no, since this graph optimizer is in a phase handling a pretty special cases (within a transformer based models), I would prefer to just support the ones we need. Does that make sense?